File: NetworkTools.js
function recommendationLoop(track)
{
    console.log("Started loop function");
    var beegArray = [];
    var token = getToken();

    let initTrack = JSON.parse(getTrack(token, track));
    beegArray.push(initTrack);

    for (let i = 0; i < 10; i++)
    {
        // I have no idea when the tokens expire so this is to avoid token timeouts
        if (i % 10 == 0)
            token = getToken();

        let rec = JSON.parse(getRecommendation(token, track));

        for (let j = 0; j < rec.tracks.length; j++)
        {
            beegArray.push(rec.tracks[j]);
        }

        track = rec.tracks[0].id;
    }

    let val = Math.floor(beegArray.length);

    var featuresRes = [];

    for (let k = 0; k < beegArray.length; k++)
    {
        if (k % 1000 == 0)
        {
            token = getToken();
        }

        if (k % 100 == 1 || k == 0)
        {
            let res = getAudioFeatures(token, beegArray[k].id);
            let resJson = JSON.parse(res)
            featuresRes.push(resJson);
        }
    }

    var currentHundred = -1;
    for (let m = 0; m < beegArray.length; m++)
    {
        if (m % 101 == 0 && currentHundred + 1 < featuresRes.length)
        {
            currentHundred++;
        }

        let obj = {
            // Lerp the values in the following format:
            // beginning + ((i / 99) * (end - beginning))
            //
            // This is very messy I know I am sorry
            acousticness: lerp(featuresRes[0+currentHundred].acousticness, featuresRes[1+currentHundred].acousticness, m%100),
            danceability: lerp(featuresRes[0+currentHundred].danceability, featuresRes[1+currentHundred].danceability, m%100),
            energy: lerp(featuresRes[0+currentHundred].energy, featuresRes[1+currentHundred].energy, m%100),
            instrumentalness: lerp(featuresRes[0+currentHundred].instrumentalness, featuresRes[1+currentHundred].instrumentalness, m%100),
            liveness: lerp(featuresRes[0+currentHundred].liveness, featuresRes[1+currentHundred].liveness, m%100),
            speechiness: lerp(featuresRes[0+currentHundred].speechiness, featuresRes[1+currentHundred].speechiness, m%100),
            valence: lerp(featuresRes[0+currentHundred].valence, featuresRes[1+currentHundred].valence, m%100),
        };

        // Append these values to the already created node
        beegArray[m] = Object.assign(beegArray[m], obj);
    }

    // Necessary to carry the values over to C++
    // VERY LARGE STRING
    let js = JSON.stringify(beegArray)

    return js;
}

function lerp(begin, end, index)
{
    // Magic linear interpolation function to make my life easier
    return begin + (((index % 100) / 99) * (end - begin));
}


function getTempo(track)
{
    let token = getToken();
    let features = JSON.parse(getAudioFeatures(token, track));
    return features.tempo.toString();
}

function getAudioFeatures(token, track)
{
    console.log("Getting features for track: " + track)
    var http = new XMLHttpRequest();
    var url = `https://api.spotify.com/v1/audio-features/${track}`;
    http.open("GET", url, false);

    http.setRequestHeader("Content-Type", "application/json");
    http.setRequestHeader("Authorization", `Bearer ${token}`);

    http.send(null);

    if (http.status === 200)
    {
        console.log("Response: " + http.responseText);
        return http.responseText;
    }
    else
    {
        return "{ msg: error }";
    }
}

function getTrack(token, track)
{
    var http = new XMLHttpRequest();
    var url = `https://api.spotify.com/v1/tracks/${track}`;
    http.open("GET", url, false);

    http.setRequestHeader("Content-Type", "application/json");
    http.setRequestHeader("Authorization", `Bearer ${token}`);

    http.send(null);

    if (http.status === 200)
    {
        console.log("Response: " + http.responseText);
        return http.responseText;
    }
    else
    {
        return "{ msg: error }";
    }
}

function getRecommendation(token, track)
{
    var http = new XMLHttpRequest();
    var url = `https://api.spotify.com/v1/recommendations?seed_tracks=${track}&limit=100`;
    http.open("GET", url, false);

    http.setRequestHeader("Content-Type", "application/json");
    http.setRequestHeader("Authorization", `Bearer ${token}`);

    http.send(null);

    if (http.status === 200)
    {
        return http.responseText;
    }
}

function getToken()
{
    var client_id = '5667d65d47f143d8922b8cfeea1f00e4';
    var client_secret = '0a1292ce57ce4c76933a223cb3ce1c88';
    var auth = Qt.btoa(client_id + ":" + client_secret);
    var fullAuth = "Basic " + auth;

    var http = new XMLHttpRequest();
    var url = "https://accounts.spotify.com/api/token";
    http.open("POST", url, false);
    http.setRequestHeader("Content-Type", "application/x-www-form-urlencoded");
    http.setRequestHeader("Authorization", fullAuth);

    http.send("grant_type=client_credentials");

    if (http.status === 200)
    {
        var js = JSON.parse(http.responseText);
        return js.access_token;
    }

}

File: networktools.cpp
#include "networktools.h"
#include "mainwindow.h"

#include <QJsonObject>
#include <QJsonDocument>

#include <QWidget>
#include <QFunctionPointer>

#include <QtNetwork/QNetworkRequest>
#include <QtNetwork/QNetworkReply>
#include <QtNetwork/QNetworkAccessManager>

#include <QList>
#include <QVariantList>

#include <QQmlEngine>
#include <QQmlComponent>
#include <QtQuick/QQuickItem>
#include <QObject>

#include <QJSValue>

#include <QtConcurrent/QtConcurrent>

// Lots of includes dang

NetworkTools::NetworkTools()
{
    qDebug() << "NetworkTools created";
}

void NetworkTools::getRequest(QString url, QVector<QString> headers)
{
    QNetworkRequest req;
    req.setUrl(url);

    for (int i = 0; i < headers.size() - 1; i += 2)
    {
        QByteArray h1 = headers.at(i).toUtf8();
        QByteArray h2 = headers.at(i + 1).toUtf8();
        req.setRawHeader(h1, h2);
    }

    qDebug() << "";
    qDebug() << "GET request req created with the following url: " << url;

    for (int i = 0; i < headers.size() - 1; i += 2)
    {
        qDebug() << "The following header: " << headers.at(i) << ": " << req.rawHeader(headers.at(i).toUtf8());
    }

    qDebug() << "";

    QNetworkReply* reply = manager.get(req);

    auto connEr = std::make_shared<QMetaObject::Connection>();
    *connEr = QObject::connect(reply, &QNetworkReply::errorOccurred, [=]() {
        qDebug() << "An error ocurred in the network request:";
        qDebug() << reply->errorString();
        qDebug() << "";
        QJsonObject jobj;
        emit getCompleted(jobj);
        QObject::disconnect(*connEr);
    });

    auto conn = std::make_shared<QMetaObject::Connection>();
    *conn = QObject::connect(reply, &QNetworkReply::finished, [=]() {
        if (reply->error() == QNetworkReply::NoError)
        {
            QJsonObject json;
            QJsonDocument doc;
            QByteArray all = reply->readAll();
            qDebug() << "Recieved response with size: " << all.size();
            qDebug().noquote() << all;

            doc = doc.fromJson(all);

            json = doc.object();

            qDebug() << "Emitting JSON:";
            qDebug() << "";
            emit getCompleted(json);
        } else
        {
            qDebug() << "An error ocurred in parsing the reply, passing an empty JSON...";
            qDebug() << reply->error();
            qDebug() << "";
            QJsonObject jobj;
            jobj.insert("Error", reply->error());
            emit getCompleted(jobj);
        }

        QObject::disconnect(*conn);
    });
}

void NetworkTools::postRequest(QString url, QVector<QString> headers, QString body)
{
    QNetworkRequest req;
    req.setUrl(url);

    for (int i = 0; i < headers.size() - 1; i += 2)
    {
        QByteArray h1 = headers.at(i).toUtf8();
        QByteArray h2 = headers.at(i + 1).toUtf8();
        req.setRawHeader(h1, h2);
    }

    qDebug() << "";
    qDebug() << "POST request req created with the following url: " << url;

    for (int i = 0; i < headers.size() - 1; i += 2)
    {
        qDebug() << "The following header: " << headers.at(i) << ": " << req.rawHeader(headers.at(i).toUtf8());
    }

    qDebug() << "And the following body: " << body.toUtf8();
    qDebug() << "";
    QNetworkReply* reply = manager.post(req, body.toUtf8());

    auto connEr = std::make_shared<QMetaObject::Connection>();
    *connEr = QObject::connect(reply, &QNetworkReply::errorOccurred, [=]() {
        qDebug() << "An error ocurred in the network request:";
        qDebug() << reply->errorString();
        qDebug() << "";
        QObject::disconnect(*connEr);
    });

    auto conn = std::make_shared<QMetaObject::Connection>();
    *conn = QObject::connect(reply, &QNetworkReply::finished, [=]() {
        qDebug() << "Server responded without error";
        if (reply->error() == QNetworkReply::NoError)
        {
            QJsonObject json;
            QJsonDocument doc;
            QByteArray all = reply->readAll();
            qDebug() << "Recieved response with size: " << all.size();
            qDebug().noquote() << all;

            doc = doc.fromJson(all);

            json = doc.object();

            qDebug() << "Emitting JSON:";
            qDebug() << "";
            emit postCompleted(json);
        } else
        {
            qDebug() << "An error ocurred in parsing the reply, passing an empty JSON...";
            qDebug() << reply->error();
            qDebug() << "";
            QJsonObject jobj;
            emit postCompleted(jobj);
        }

        QObject::disconnect(*conn);
    });
}

void NetworkTools::getToken()
{
    // Generate the authentication information for the header
    // NOTE: these probably shouldn't be constants in the program, but it works for now
    QString client_id = "5667d65d47f143d8922b8cfeea1f00e4";
    QString client_secret = "0a1292ce57ce4c76933a223cb3ce1c88";
    QString auth = client_id + ":" + client_secret;
    auth = "Basic " + auth.toUtf8().toBase64();

    QVector<QString> headers;
    headers.push_back("Content-Type");
    headers.push_back("application/x-www-form-urlencoded");
    headers.push_back("Authorization");
    headers.push_back(auth);

    QJsonObject result;

    auto conn = std::make_shared<QMetaObject::Connection>();
    *conn = QObject::connect(this, &NetworkTools::postCompleted, [=](QJsonObject obj) {
        QString token = obj.find("access_token").value().toString();

        emit tokenFound(token);
        QObject::disconnect(*conn);
    });

    postRequest(
                "https://accounts.spotify.com/api/token",
                headers,
                "grant_type=client_credentials");

}

void NetworkTools::getToken(MainWindow* parent)
{
    // Generate the authentication information for the header
    // NOTE: these probably shouldn't be constants in the program, but it works for now
    QString client_id = "5667d65d47f143d8922b8cfeea1f00e4";
    QString client_secret = "0a1292ce57ce4c76933a223cb3ce1c88";
    QString auth = client_id + ":" + client_secret;
    auth = "Basic " + auth.toUtf8().toBase64();

    QVector<QString> headers;
    headers.push_back("Content-Type");
    headers.push_back("application/x-www-form-urlencoded");
    headers.push_back("Authorization");
    headers.push_back(auth);

    QJsonObject result;

    auto conn = std::make_shared<QMetaObject::Connection>();
    *conn = QObject::connect(this, &NetworkTools::postCompleted, [=](QJsonObject obj) {
        QString token = obj.find("access_token").value().toString();

        emit parent->tokenRequestCompleted(token);
        QObject::disconnect(*conn);
    });

    postRequest(
                "https://accounts.spotify.com/api/token",
                headers,
                "grant_type=client_credentials");

}

void NetworkTools::searchSong(QString token, QString query, MainWindow* parent)
{
    QVector<QString> vec;
    vec.push_back("Content-Type");
    vec.push_back("application/json");

    QString auth = "Bearer " + token;
    vec.push_back("Authorization");
    vec.push_back(auth);

    QString url = "https://api.spotify.com/v1/search?q=" + query + "&type=track";

    auto conn = std::make_shared<QMetaObject::Connection>();
    *conn = QObject::connect(this, &NetworkTools::getCompleted, [=](QJsonObject obj) {
        emit parent->searchResultCompleted(obj);
        QObject::disconnect(*conn);
    });

    getRequest(url, vec);
}

void NetworkTools::getSongInfo(QString token, QString songName, MainWindow *parent)
{
    QVector<QString> vec;
    vec.push_back("Content-Type");
    vec.push_back("application/json");

    QString auth = "Bearer " + token;
    vec.push_back("Authorization");
    vec.push_back(auth);

    QString url = "https://api.spotify.com/v1/audio-features/" + songName;

    auto conn = std::make_shared<QMetaObject::Connection>();
    *conn = QObject::connect(this, &NetworkTools::getCompleted, [=](QJsonObject obj) {
        emit parent->songResultCompleted(obj);
        // Disconnect lambda
        QObject::disconnect(*conn);
    });

    getRequest(url, vec);
}



void NetworkTools::recommendationsLoop(QString songId, MainWindow *parent)
{
    QQmlEngine engine;

    QQmlComponent component(&engine, QUrl("qrc:main/dummy.qml"), QQmlComponent::PreferSynchronous);

    qDebug() << component.errorString();
    if (component.status() == QQmlComponent::Ready)
    {
        QObject *object = component.create();

        QString list;
        QVariant returned;

        // Do not run a synchronous network request on the UI thread, you say?
        // Nay I say, nay
        // This *will* freeze the process
        QMetaObject::invokeMethod(object, "doRecLoop", Q_RETURN_ARG(QVariant, returned), Q_ARG(QString, songId));

        list = returned.toString();

        QJsonDocument doc = QJsonDocument::fromJson(list.toUtf8());
        QJsonArray arr = doc.array();
        emit (parent->recommendedCompleted(arr));
    }
}

QString NetworkTools::getTempo(QString songId)
{
    QQmlEngine engine;

    QQmlComponent component(&engine, QUrl("qrc:main/dummy.qml"), QQmlComponent::PreferSynchronous);

    if (component.status() == QQmlComponent::Ready)
    {
        QObject *object = component.create();

        QString tempo;
        QVariant returned;
        QVariant id = QVariant::fromValue(songId);

        // Do not run a synchronous network request on the UI thread, you say?
        // Nay I say, nay (again)
        // This *will* freeze the process (much less time than the other one tho)
        QMetaObject::invokeMethod(object, "getTempo", Q_RETURN_ARG(QVariant, returned), Q_ARG(QVariant, id));

        tempo = returned.toString();

        return tempo;
    }
    else
    {
        return "";
    }
}
